source/common/thread_local/thread_local_impl.cc

```diff
@@ -93,24 +112,24 @@ void InstanceImpl::setThreadLocal(uint32_t index, ThreadLocalObjectSharedPtr obj
 
 void InstanceImpl::shutdownGlobalThreading() {
   ASSERT(std::this_thread::get_id() == main_thread_id_);
+  ASSERT(!shutdown_);
   shutdown_ = true;
 }
 
 void InstanceImpl::shutdownThread() {
+  ASSERT(shutdown_);
+
   // Destruction of slots is done in *reverse* order. This is so that filters and higher layer
   // things that are built on top of the cluster manager, stats, etc. will be destroyed before
   // more base layer things. It's possible this might need to become more complicated later but
-  // it's OK for now.
+  // it's OK for now. Note that this is always safe to do because:
```

Comment:
I'm still wondering why it's not possible pre-shutdown for there to have been a sequence of slot creations and removals that leads to reverse ordering not reflecting object dependency. E.g. object X has suboject Y, both of which allocate a slot. However, object Z allocates before X, removes its slot after X's allocation and Y's allocation has Z's old slot? 

The post seems to make this safe, but then I'm wondering why we bother with reverse ordering in the first place. I'm probably missing something, sorry.

Other than this, LGTM.